### PR TITLE
[WIP] Add pytorch support through a wrapper function

### PR DESCRIPTION
**Pull request by nottombrown**
Preliminary PR that lets people apply `cleverhans` attacks to custom pytorch models by wrapping the whole pytorch model as a custom tf op that passes the gradient through using chain rule.

I see two main complaints that people might have:

1. This still requires constructing a static graph for eval, which may make it unattractive for die-hard pytorch users
2. The `pytorch => tf` wrapping can make debugging quite difficult

@david-berthelot, interested in your feedback on this preliminary pull-request because you're the other main pytorch user on the team.

### CODE REVIEW

File: cleverhans/utils_pytorch.py [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+    """Convert a pytorch model into a tensorflow op that allows backprop"""
+    torch_state = {
+        'logits': None,
+        'x': None,
+    }
+    if not out_features:
+        out_features = list(model.modules())[-1].out_features
+
+    def _fprop_fn(x_np):
+        x_tensor = torch.Tensor(x_np).cuda()
```
Comment (david-berthelot) [inline cleverhans/utils_pytorch.py]: I think you forgot to remove .cuda() here.

File: cleverhans/utils_pytorch.py [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+    g = tf.get_default_graph()
+
+    # Add gradient override map
+    with g.gradient_override_map(
+            {"PyFunc": rnd_name, "PyFuncStateless": rnd_name}):
+        return tf.py_func(func, inp, Tout, stateful=stateful, name=name)
+
+
+def convert_pytorch_model_to_tf(model, out_features=None):
+    """Convert a pytorch model into a tensorflow op that allows backprop"""
```
Comment (david-berthelot) [inline cleverhans/utils_pytorch.py]: Please document the :param args.

File: cleverhans/utils_pytorch.py [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+                          Tout=[tf.float32])
+
+    def tf_model_op(x_op):
+        out = _py_func_with_gradient(_fprop_fn, [x_op], Tout=[tf.float32],
+                                     stateful=True,
+                                     grad_func=_tf_gradient_fn)[0]
+        out.set_shape([None, out_features])
+        return out
+
+    return tf_model_op
```
Comment (david-berthelot) [inline cleverhans/utils_pytorch.py]: I'm impressed at how elegant and readable your implementation is.

File: cleverhans/utils_pytorch.py (hunk 1 above)
Comment (nottombrown) [reply cleverhans/utils_pytorch.py]: Good call!

File: cleverhans/utils_pytorch.py (hunk 2 above)
Comment (nottombrown) [reply cleverhans/utils_pytorch.py]: Updated

File: cleverhans/utils_pytorch.py (hunk 3 above)
Comment (nottombrown) [reply cleverhans/utils_pytorch.py]: Thanks!
